app/code/Magento/CatalogUrlRewrite/Model/Product/GetProductUrlRewriteDataByStore.php

```diff
@@ -0,0 +1,66 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\CatalogUrlRewrite\Model\Product;
+
+use Magento\Catalog\Api\Data\ProductInterface;
+use Magento\CatalogUrlRewrite\Model\ResourceModel\Product\GetUrlRewriteData;
+use Magento\Store\Model\Store;
+
+/**
+ * Product data needed for url rewrite generation locator class
+ */
+class GetProductUrlRewriteDataByStore
+{
+    /**
+     * @var array
+     */
+    private $urlRewriteData = [];
+
+    /**
+     * @var GetUrlRewriteData
+     */
+    private $getUrlRewriteData;
+
+    /**
+     * @param GetUrlRewriteData $getUrlRewriteData
+     */
+    public function __construct(GetUrlRewriteData $getUrlRewriteData)
+    {
+        $this->getUrlRewriteData = $getUrlRewriteData;
+    }
+
+    /**
+     * Retrieves data for product by store
+     *
+     * @param ProductInterface $product
+     * @param int $storeId
+     * @return array
+     */
+    public function execute(ProductInterface $product, int $storeId): array
+    {
+        $productId = $product->getId();
+        if (isset($this->urlRewriteData[$productId][$storeId])) {
+            return $this->urlRewriteData[$productId][$storeId];
```

Comment:
Its a little bit dangerous to have cache here. You should at least have method to reset cache and do it after rewrites changed

Reply:
added clearing of local cache by product after generation

Reply:
Please name it **clearProductUrlRewriteDataCache** to clarify that we clear exactly cache, not url rewrite data itself

Reply:
done